### PR TITLE
PR5 - Audio - Spectrogram HOST Tensor

### PR DESCRIPTION
**Pull request by snehaa8**
This is the updated spectrogram kernel with all latest changes

### CODE REVIEW

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+*/
+
+#include "rppdefs.h"
+#include "../../../../third_party/ffts/include/ffts.h"
+#include "../../../../third_party/ffts/include/ffts_attributes.h"
```
Comment (r-abishek) [inline src/modules/cpu/kernel/spectrogram.hpp:25]: Suggest changing the folder structure to rpp/include/third_party/ffts and placing the 2 includes there.
And change this line to:
```
#include "third_party/ffts/ffts.h"
#include "third_party/ffts/ffts_attributes.h"
```

File: src/modules/rppt_tensor_audio_augmentations.cpp [hunk 2]
```diff
@@ -124,3 +124,75 @@ RppStatus rppt_pre_emphasis_filter_host(RppPtr_t srcPtr,
...
+        return RPP_ERROR_NOT_IMPLEMENTED;
+    }
+}
+
+/******************** spectrogram ********************/
+
+RppStatus rppt_spectrogram_host(RppPtr_t srcPtr,
+                                RpptDescPtr srcDescPtr,
+                                RppPtr_t dstPtr,
+								RpptDescPtr dstDescPtr,
```
Comment (r-abishek) [inline src/modules/rppt_tensor_audio_augmentations.cpp:163]: Convert tabs to spaces on L163 in vscode

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+    const Rpp32f mulFactor = (2.0 * M_PI) / nfft;
+    const Rpp32u hStride = dstDescPtr->strides.hStride;
+    const Rpp32s alignedNfftLength = nfft & ~7;
+    const Rpp32s alignedNbinsLength = numBins & ~7;
+    const Rpp32s alignedWindowLength = windowLength & ~7;
+    bool useRealImpl = can_use_real_impl(nfft);
+    const auto fftInSize = size_in_buf(nfft);
+    const auto fftOutSize = size_out_buf(nfft);
+
+    Rpp32f *windowFn = static_cast<Rpp32f *>(calloc(windowLength, sizeof(Rpp32f)));
```
Comment (r-abishek) [inline src/modules/cpu/kernel/spectrogram.hpp:99]: free?

File: src/modules/cpu/kernel/spectrogram.hpp [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+        // Generate FFT output
+        ffts_plan_t *p;
+        if(useRealImpl)
+            p = ffts_init_1d_real(nfft, FFTS_FORWARD);
+        else
+            p = ffts_init_1d(nfft, FFTS_FORWARD);
+
+        if (!p)
+        {
+            printf("FFT Plan is unsupported. Exiting the code\n");
```
Comment (r-abishek) [inline src/modules/cpu/kernel/spectrogram.hpp:173]: Remove print statement. Create an error code RPP_ERROR_FFT_PLAN_UNSUPPORTED and return that. In the cpp API file, return whats returned here.

File: utilities/test_suite/HOST/Tensor_host_audio.cpp [hunk 5]
```diff
@@ -195,6 +195,78 @@ int main(int argc, char **argv)
...
+                            maxDstWidth = std::max(maxDstWidth, (int)dstDims[i].width);
+                        }
+                    }
+
+                    dstDescPtr->w = maxDstWidth;
+                    dstDescPtr->h = maxDstHeight;
+
+                    dstDescPtr->strides.nStride = dstDescPtr->c * dstDescPtr->w * dstDescPtr->h;
+                    dstDescPtr->strides.hStride = dstDescPtr->c * dstDescPtr->w;
+                    dstDescPtr->strides.wStride = dstDescPtr->c;
```
Comment (r-abishek) [inline utilities/test_suite/HOST/Tensor_host_audio.cpp:258]: Cant we use the set_audio_descriptor_dims_and_strides() for this?

File: CMakeLists.txt [hunk 6]
```diff
@@ -296,6 +296,7 @@ message("-- ${White}${PROJECT_NAME} -- Link Libraries: ${LINK_LIBRARY_LIST}${Col
 target_link_libraries(${PROJECT_NAME} ${LINK_LIBRARY_LIST})
 set_target_properties(${PROJECT_NAME} PROPERTIES POSITION_INDEPENDENT_CODE ON)
 set_target_properties(${PROJECT_NAME} PROPERTIES LINKER_LANGUAGE CXX)
+target_link_libraries(${PROJECT_NAME} ${PROJECT_SOURCE_DIR}/third_party/ffts/src/libffts.a)
```
Comment (r-abishek) [inline CMakeLists.txt:299]: For the .a file, suggest creating the folder structure - rpp/libs/third_party/ffts/libffts.a

**Review by r-abishek — CHANGES_REQUESTED** [review]
Please do a ToT merge and address review comments

File: include/rppdefs.h [hunk 7]
```diff
@@ -396,6 +396,15 @@ typedef enum
     REFLECT
 } RpptAudioBorderType;
 
+/*! \brief RPPT Spectrogram Layout enum
+ * \ingroup group_rppdefs
+ */
+typedef enum
+{
+    FT = 0,
+    TF,
```
[the remark this answers is not in the document]
Comment (snehaa8) [reply include/rppdefs.h:405]: Done

File: src/modules/cpu/kernel/spectrogram.hpp (hunk 4 above)
Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:173]: We won't be able to judge anything on ffts_plan_t in cpp API file, please correct me if am wrong.
p is based on 
```
if(useRealImpl)
            p = ffts_init_1d_real(nfft, FFTS_FORWARD);
        else
            p = ffts_init_1d(nfft, FFTS_FORWARD);
```

Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:173]: Removed print statement and returned error code RPP_ERROR_FFT_PLAN_UNSUPPORTED

Comment (snehaa8) [reply src/modules/cpu/kernel/spectrogram.hpp:173]: Adding return inside openMP segment leads to error.

Comment (r-abishek) [reply src/modules/cpu/kernel/spectrogram.hpp:175]: @snehaa8 We ideally don't want print statements inside lib code. Lets find a work around, and you can issue a separate PR.

**Review by r-abishek — APPROVED** [review]
Moving ahead for the external PR to validate on CI. Clarifications internal/external on separate discussion.

**Comment by r-abishek** [comment]
Develop merge + conflicts fixed

**Pull request merged**